main/src/ca/uwaterloo/flix/api/lsp/StackConsumer.scala

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2024 Alexander Dybdahl Troelsen
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package ca.uwaterloo.flix.api.lsp
+
+import ca.uwaterloo.flix.api.lsp.Visitor.Consumer
+import ca.uwaterloo.flix.language.ast.Ast.{AssocTypeConstructor, Derivation, Derivations, EqualityConstraint, TraitConstraint}
+import ca.uwaterloo.flix.language.ast.Type
+import ca.uwaterloo.flix.language.ast.TypedAst.Pattern.Record.RecordLabelPattern
+import ca.uwaterloo.flix.language.ast.TypedAst.*
+import ca.uwaterloo.flix.language.ast.shared.Annotation
+import ca.uwaterloo.flix.language.ast.shared.SymUse.*
+
+/**
+  * Consumer that collects every visited AST nodes on a stack where the head is the last element visited.
+  */
+case class StackConsumer() extends Consumer {
+  private var stack: List[AnyRef] = Nil
+
+  private def push(x: AnyRef): Unit = {
+    stack = x :: stack
+  }
+
+  /**
+    * @return [[Option]] containing the top element from the stack of AST elements that the consumer has collected.
+    *         Returns [[Null]] if the stack is empty.
+    */
+  def getHead: Option[AnyRef] = stack.headOption
+
+  /**
+    * @return The stack of AST elements that the consumer has collected.
+    */
+  def getStack: List[AnyRef] = stack
```

Comment:
I suggest to keep getStack, but to drop getHead and getTail. You have direct access to those operations on the list anyway.

Reply:
true

Reply:
i suppose those only really make sense if you specifically didn't want to expose the entire list (for some reason)

Reply:
Yes, so kill them and then I can merge.

Reply:
yep